added alternative solution for CSES-1670 (#3839)
cpinitiative/usaco-guide: solutions/gold/cses-1670.mdx

id: cses-1670
source: CSES
title: Swap Game
author: Maggie Liu, Daniel Ge
---

<Spoiler title="Hint">
Think of all $9!$ possible configurations of the board as the nodes, and our moves
as the edges that connect them. We then perform BFS, starting at our intial given

> configuration.

Reviewer: Think of all $9!$ possible configurations of the board as the nodes with the moves
as the edges that connect them.

</Spoiler>

> The difficulty of this problem is with how we choose to implement the BFS. We present

Reviewer 2: the difference is the way the state is hashed, right?

Author: yes but I didn't want to use that word since it's a later unit

Reviewer 3: just mention & link the future module :>

Author: sure!

> two solutions: Solution 1 runs faster, and solution 2 is easier to write (but may

Reviewer: ```suggestion
two solutions with the first being faster and the second being easier to write (note that the second solution may TLE)
```

```

</CPPSection>
</LanguageSection>
</Spoiler>

<Spoiler title="Solution 2">
<Warning>
While this solution is much easier to implement, it is extremely difficult to write an implementation that does not TLE in Java or Python. It may also even TLE in C++ if written inefficiently.
</Warning>
## Explanation
Since there are only $9!$ states the grid can be in, we can use BFS to find the
number of moves needed to reach the $\texttt{target}$ from the starting grid. We
represent each grid as a string read left-to-right then top-to-bottom.

For example, the grid in the testcase,
```
2 1 3
7 5 9
8 4 6
```
becomes "213759846".

We then perform BFS normally but with a few notable changes. First, since we can't have
strings be the indices of an array, we store the "depth" of each string (number of moves required to reach a board configuration) inside our queue.

We also store the strings (board configurations) already seen inside an **unordered_set**, as using normal (ordered) sets would be too slow.
<Info title="Significance of using unordered set">

> As we will learn in the [hashing module](/gold/hashmaps), unordered set does searches and insertions in constant time despite a large constant factor while a normal set does this in $\mathcal{O}(\log n)$. Since we are dealing with up to $9!$ terms, it is necessary to use unordered set.

Reviewer 3: ```suggestion
As we will learn in the [hashing module](/gold/hashmaps), unordered set does searches and insertions in constant time despite a large constant factor while a normal set does this in $\mathcal{O}(\log n)$. Since we are dealing with up to $9!$ terms, it is necessary to use an unordered set.
```

</Info>

## Implementation

**Time Complexity:** $\mathcal{O}(N^2!)$, where $N$ is the side length of the
grid

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

const int fact[9] = {1, 1, 2, 6, 24, 120, 720, 5040, 40320}, fact9 = 362880;
unordered_set<string> visited;

// Defined globally to be used in processwap
queue<pair<string, int>> q;
int moves;
string curboard;

// Processing swapping the numbers in x, y positions

> void processwap(int x, int y) {

Reviewer 3: rename to `process_swap`

```

</CPPSection>
</LanguageSection>
</Spoiler>

<Spoiler title="Solution 2">
<Warning>
While this solution is much easier to implement, it is extremely difficult to write an implementation that does not TLE in Java or Python. It may also even TLE in C++ if written inefficiently.

> </Warning>

Reviewer: Why is this solution good then? If it doesn't easily pass in C++/Java, I'm not sure it's worth adding

Author: There are many CSES (and even USACO) problems that TLEs using Java/Python or an inefficient C++ code even with the right approach. (e.g. https://cses.fi/problemset/task/1635/)
![image](https://github.com/cpinitiative/usaco-guide/assets/52391257/87cb61a8-3e09-413b-9e42-adc3344dd524)
(this code runs the biggest test case in ~1300ms, and I had to do a small optimization by using constant variables)

In terms of educational value, it also encourages readers to practice coefficient optimization in my opinion. The approach I gave was to use unordered_set, but there more

Reviewer 3: if this is an intended solution, then it's fine the way it is

if not, maybe replace title with "Slower Solution"

Reviewer 3: ```suggestion
<Warning>
While this solution is much easier to implement, it is extremely difficult to write an implementation that does not TLE in Java or Python. It may TLE in C++ if written inefficiently.
</Warning>
```

## Explanation
Since there are only $9!$ states the grid can be in, we can use BFS to find the
number of moves needed to reach the $\texttt{target}$ from the starting grid. We
represent each grid as a string read left-to-right then top-to-bottom.

For example, the grid in the testcase,
```
2 1 3
7 5 9
8 4 6
```
becomes "213759846".

> We then perform BFS normally but with a few notable changes. Since we cannot have
> our strings be the indices of an array, we store the "depth" of each string (number of moves required to reach a board configuration) inside our queue.

Reviewer: ```suggestion
We then perform BFS normally but with a few notable changes. First, since we can't have
strings be the indices of an array, we store the "depth" of each string (number of moves required to reach a board configuration) inside our queue.
```

> In addition, we store the strings (board configurations) we have already seen inside an **unordered_set**, as using normal (ordered) sets will TLE.

Reviewer: ```suggestion
We also store the strings (board configurations) already seen inside an **unordered_set**, as using normal (ordered) sets would be too slow.
```

<Info title="Significance of using unordered set">

> As we will learn in the [hashing module](/gold/hashmaps), unordered set does searching and inserting in $\mathcal{O}(1)$ albiet with a large constant factor. A normal set does this in $\mathcal{O}(\log n)$, so since we are dealing with up to $9!$ terms, it is necessary to use unordered set.

Reviewer: ```suggestion
As we will learn in the [hashing module](/gold/hashmaps), unordered set does searches and insertions in constant time despite a large constant factor while a normal set does this in $\mathcal{O}(\log n)$. Since we are dealing with up to $9!$ terms, it is necessary to use unordered set.
```

</Info>

## Implementation

**Time Complexity:** $\mathcal{O}(N^2!)$, where $N$ is the side length of the
grid

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

> const int fact[9] = {1, 1, 2, 6, 24, 120, 720, 5040, 40320}, fact9 = 362880;

Reviewer 3: not fixed?

> const int fact[9] = {1, 1, 2, 6, 24, 120, 720, 5040, 40320}, fact9 = 362880;

Author: oops mb

```

</CPPSection>
</LanguageSection>
</Spoiler>

<Spoiler title="Solution 2">
<Warning>
While this solution is much easier to implement, it is extremely difficult to write an implementation that does not TLE in Java or Python. It may TLE in C++ if written inefficiently.
</Warning>
## Explanation
Since there are only $9!$ states the grid can be in, we can use BFS to find the
number of moves needed to reach the $\texttt{target}$ from the starting grid. We
represent each grid as a string read left-to-right then top-to-bottom.

For example, the grid in the testcase,
```
2 1 3
7 5 9
8 4 6
```
becomes "213759846".

> We then perform BFS normally but with a few notable changes. First, since we can't have
> strings be the indices of an array, we store the "depth" of each string (number of moves required to reach a board configuration) inside our queue.
> 
> We also store the strings (board configurations) already seen inside an **unordered_set**, as using normal (ordered) sets would be too slow.
> <Info title="Significance of using unordered set">
> As we will learn in the [hashing module](/gold/hashmaps), unordered set does searches and insertions in constant time despite a large constant factor while a normal set does this in $\mathcal{O}(\log n)$. Since we are dealing with up to $9!$ terms, it is necessary to use an unordered set.

Reviewer 3: ```suggestion
We then perform BFS normally but with a few notable changes.

1. Since we can't have strings be the indices of an array, we store the "depth" of each string (number of moves required to reach a board configuration) inside our queue.

2. We also store the strings (board configurations) already seen inside an **unordered_set**, as using normal (ordered) sets would be too slow.

<Info title="Significance of using unordered set">
As we will learn in the [hashing module](/gold/hashmaps), unordered set does searches and insertions in constant time despite a large constant factor while a normal set does this in $\mathcal{O}(\log n)$. Since we are dealing with up to $9!$ terms, it is necessary to use an unordered set.
```
